hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/common/AbstractNativeSystemContract.java

```diff
@@ -0,0 +1,189 @@
+                    enhancement
+                            .systemOperations()
+                            .externalizeResult(
+                                    successResultOf(
+                                            attempt.senderId(),
+                                            pricedResult.fullResult(),
+                                            frame,
+                                            !call.allowsStaticFrame()),
+                                    pricedResult.responseCode(),
+                                    enhancement
+                                            .systemOperations()
+                                            .syntheticTransactionForNativeCall(input, contractID, true));
+                } else {
+                    externalizeFailure(
+                            gasRequirement,
+                            input,
+                            insufficientGas
+                                    ? Bytes.EMPTY
+                                    : pricedResult.fullResult().output(),
+                            attempt,
+                            status,
+                            enhancement,
+                            contractID);
+                }
+            }
+        } catch (final HandleException handleException) {
+            return haltHandleException(handleException, frame.getRemainingGas());
+        } catch (final Exception internal) {
+            log.error("Unhandled failure for input {} to native system contract", input, internal);
+            return haltResult(PRECOMPILE_ERROR, frame.getRemainingGas());
+        }
+        return pricedResult.fullResult();
+    }
+
+    private static void externalizeFailure(
+            final long gasRequirement,
+            @NonNull final Bytes input,
+            @NonNull final Bytes output,
+            @NonNull final AbstractCallAttempt attempt,
+            @NonNull final ResponseCodeEnum status,
+            @NonNull final HederaWorldUpdater.Enhancement enhancement,
+            @NonNull final ContractID contractID) {
+        enhancement
+                .systemOperations()
+                .externalizeResult(
+                        contractFunctionResultFailedFor(
+                                attempt.senderId(), output, gasRequirement, status.toString(), contractID),
+                        status,
+                        enhancement.systemOperations().syntheticTransactionForNativeCall(input, contractID, true));
+    }
+
+    // potentially other cases could be handled here if necessary
+    private static FullResult haltHandleException(final HandleException handleException, long remainingGas) {
+        if (handleException.getStatus().equals(MAX_CHILD_RECORDS_EXCEEDED)) {
+            return haltResult(CustomExceptionalHaltReason.INSUFFICIENT_CHILD_RECORDS, remainingGas);
+        }
+        throw handleException;
+    }
+
+    //
+    protected abstract FrameUtils.CallType callTypeOf(final MessageFrame frame);
```

Comment:
Annotate arg `frame` as `@NonNull`; javadoc (since abstract) - or is it just obvious given arg type and result type?